### PR TITLE
[fix](binlog) Refresh incremental partition versions before pruning

### PR DESCRIPTION
**Pull request by luwei16**
### What problem does this PR solve?

Issue Number: None

Related PR: #67181

Problem Summary:

A cloud incremental read may wait for its target transaction to become visible before query planning continues. During that wait, the FE `CloudPartition` cache can still contain `PARTITION_INIT_VERSION` for a partition even though the transaction has already become visible in MetaService.

`PruneEmptyPartition` previously relied only on that cached version. It could therefore classify the partition as empty and remove it before scan-node planning. The later version refresh in `ScanNode.setVisibleVersionForOlapScanNodes` could observe the new version, but it was already too late because the partition had been pruned from the scan.

This PR changes empty-partition pruning for direct cloud incremental reads as follows:

- A partition that is already known to be non-empty from the FE cache is retained directly. Partition versions are monotonic, so a cached version greater than `PARTITION_INIT_VERSION` remains sufficient for this pruning decision.
- Only cached-empty or unknown candidate partitions are refreshed from MetaService. This catches transactions that became visible after the FE cache entry was populated while avoiding unnecessary RPC work for known non-empty partitions.
- The refreshed and cached results are combined in the original candidate-partition order.
- Normal reads and incremental reads backed by an `OlapTableWrapper` with fixed visible versions continue to use their existing snapshot-aware path.

The refresh in `ScanNode.setVisibleVersionForOlapScanNodes` is intentionally retained. `PruneEmptyPartition` only needs an empty-or-non-empty decision, while scan execution still needs the exact snapshot versions of all final selected partitions.

### Release note

Fix cloud incremental reads that could miss newly visible data because a stale FE partition-version cache caused premature empty-partition pruning.

### Check List (For Author)

- Test
    - [ ] Regression test
    - [x] Unit Test
        - `PruneEmptyPartitionTest`
        - `OlapScanNodeTest`
    - [ ] Manual test (add detailed scripts or steps below)
    - [ ] No need to test or manual test. Explain why:
        - [ ] This is a refactor/code format and no logic has been changed.
        - [ ] Previous test can cover this change.
        - [ ] No code files have been changed.
        - [ ] Other reason

- Behavior changed:
    - [ ] No.
    - [x] Yes. Direct cloud incremental reads refresh cached-empty or unknown partition versions from MetaService before empty-partition pruning.

- Does this need documentation?
    - [x] No.
    - [ ] Yes.

### Check List (For Reviewer who merge this PR)

- [ ] Confirm the release note
- [ ] Confirm test cases
- [ ] Confirm document
- [ ] Add branch pick label

### CODE REVIEW

**Comment by hello-stephen** [comment]
Thank you for your contribution to Apache Doris.
Don't know what should be done next? See [How to process your PR](https://doris.apache.org/community/how-to-contribute/review-and-ci).

Please clearly describe your PR:
1. What problem was fixed (it's best to include specific error reporting information). How it was fixed.
2. Which behaviors were modified. What was the previous behavior, what is it now, why was it modified, and what possible impacts might there be.
3. What features were added. Why was this function added?
4. Which code was refactored and why was this part of the code refactored?
5. Which functions were optimized and what is the difference before and after the optimization?

**Comment by luwei16** [comment]
run buildall

**Comment by hello-stephen** [comment]
<details>
<summary>TPC-H: <b>Total hot run time: 27678 ms</b></summary>

```
machine: 'aliyun_ecs.c7a.8xlarge_32C64G'
scripts: https://github.com/apache/doris/tree/master/tools/tpch-tools
Tpch sf100 test result on commit e192f752ecf3f665bc298dd5103aadfcedb65267, data reload: false

------ Round 1 ----------------------------------
============================================
q1	17608	3926	3865	3865
q2	2320	376	299	299
q3	10005	1399	796	796
q4	4681	481	348	348
q5	7470	814	538	538
q6	181	171	136	136
q7	719	797	574	574
q8	9361	1577	1586	1577
q9	5481	4224	4196	4196
q10	6756	1610	1324	1324
q11	437	273	238	238
q12	637	416	289	289
q13	18034	2610	2014	2014
q14	256	248	230	230
q15	q16	719	712	665	665
q17	1725	1164	971	971
q18	6426	5561	5529	5529
q19	1166	1186	976	976
q20	473	397	262	262
q21	5568	2832	2561	2561
q22	438	373	290	290
Total cold run time: 100461 ms
Total hot run time: 27678 ms

----- Round 2, with runtime_filter_mode=off -----
============================================
q1	4629	4488	4327	4327
q2	819	605	545	545
q3	4869	5422	4574	4574
q4	2241	2353	1466	1466
q5	4760	4478	4546	4478
q6	233	181	132	132
q7	1842	1696	1505	1505
q8	2359	2266	1926	1926
q9	7517	7179	6851	6851
q10	4210	4200	3822	3822
q11	522	375	335	335
q12	683	717	509	509
q13	2293	2604	1998	1998
q14	259	272	249	249
q15	q16	664	677	593	593
q17	7253	6722	6653	6653
q18	11873	11072	11678	11072
q19	1102	1005	981	981
q20	2210	2202	1912	1912
q21	5306	4414	4633	4414
q22	502	472	396	396
Total cold run time: 66146 ms
Total hot run time: 58738 ms
```
</details>

**Comment by hello-stephen** [comment]
<details>
<summary>TPC-DS: <b>Total hot run time: 153152 ms</b></summary>

```
machine: 'aliyun_ecs.c7a.8xlarge_32C64G'
scripts: https://github.com/apache/doris/tree/master/tools/tpcds-tools
TPC-DS sf100 test result on commit e192f752ecf3f665bc298dd5103aadfcedb65267, data reload: false

query5	4344	595	439	439
query6	454	186	174	174
query7	4800	530	284	284
query8	315	168	166	166
query9	8780	3916	3906	3906
query10	446	305	256	256
query11	5843	2166	1991	1991
query12	158	94	95	94
query13	1337	579	415	415
query14	6457	4537	4213	4213
query14_1	4072	3978	3976	3976
query15	202	197	177	177
query16	966	463	447	447
query17	911	708	558	558
query18	2435	442	333	333
query19	199	177	139	139
query20	104	101	101	101
query21	220	134	117	117
query22	13050	13163	12845	12845
query23	15450	14463	13942	13942
query23_1	14012	14039	14107	14039
query24	7468	1683	1205	1205
query24_1	1267	1204	1224	1204
query25	536	436	369	369
query26	1306	327	167	167
query27	2625	586	321	321
query28	4522	1964	1969	1964
query29	1032	599	479	479
query30	316	241	204	204
query31	896	755	633	633
query32	143	97	97	97
query33	508	333	252	252
query34	1200	1097	624	624
query35	720	750	658	658
query36	791	786	719	719
query37	170	107	92	92
query38	1839	1772	1682	1682
query39	700	699	695	695
query39_1	652	649	629	629
query40	226	120	103	103
query41	71	70	69	69
query42	99	89	89	89
query43	336	343	305	305
query44	1344	695	705	695
query45	190	177	174	174
query46	1051	1146	732	732
query47	1489	1497	1377	1377
query48	402	399	294	294
query49	598	424	295	295
query50	971	347	247	247
query51	10558	10633	10355	10355
query52	84	86	73	73
query53	231	245	177	177
query54	256	195	205	195
query55	75	74	67	67
query56	242	218	201	201
query57	1471	1387	1349	1349
query58	235	210	212	210
query59	1976	2038	1809	1809
query60	277	242	217	217
query61	136	142	142	142
query62	414	317	269	269
query63	221	173	172	172
query64	2653	1001	848	848
query65	4014	3962	3928	3928
query66	1781	429	308	308
query67	19946	19914	19911	19911
query68	3073	1528	900	900
query69	394	289	266	266
query70	949	885	866	866
query71	290	231	208	208
query72	2866	2424	2162	2162
query73	804	760	421	421
query74	4600	4484	4248	4248
query75	2302	2256	1904	1904
query76	2311	1074	681	681
query77	347	401	293	293
query78	8987	9112	8423	8423
query79	1341	1213	749	749
query80	560	442	370	370
query81	466	279	238	238
query82	624	153	119	119
query83	367	269	251	251
query84	319	151	116	116
query85	874	465	369	369
query86	348	226	219	219
query87	1971	1960	1843	1843
query88	3655	2695	2680	2680
query89	361	283	246	246
query90	1881	175	176	175
query91	167	156	125	125
query92	94	92	78	78
query93	1573	1417	932	932
query94	531	337	306	306
query95	639	358	416	358
query96	1010	754	343	343
query97	2440	2456	2298	2298
query98	199	190	186	186
query99	725	726	608	608
Total cold run time: 239850 ms
Total hot run time: 153152 ms
```
</details>

**Comment by hello-stephen** [comment]
<details>
<summary>ClickBench: <b>Total hot run time: 23.94 s</b></summary>

```
machine: 'aliyun_ecs.c7a.8xlarge_32C64G'
scripts: https://github.com/apache/doris/tree/master/tools/clickbench-tools
ClickBench test result on commit e192f752ecf3f665bc298dd5103aadfcedb65267, data reload: false

query1	0.01	0.01	0.01
query2	0.09	0.05	0.05
query3	0.26	0.14	0.13
query4	1.62	0.14	0.14
query5	0.24	0.23	0.22
query6	1.15	0.92	0.91
query7	0.04	0.01	0.01
query8	0.06	0.04	0.03
query9	0.39	0.34	0.34
query10	0.55	0.54	0.55
query11	0.20	0.14	0.14
query12	0.18	0.15	0.16
query13	0.46	0.47	0.47
query14	0.96	0.94	0.94
query15	0.60	0.60	0.57
query16	0.32	0.32	0.31
query17	1.09	1.05	1.10
query18	0.22	0.20	0.21
query19	2.01	1.90	1.94
query20	0.02	0.01	0.02
query21	15.42	0.22	0.15
query22	4.74	0.05	0.05
query23	16.12	0.30	0.13
query24	3.02	0.41	0.33
query25	0.13	0.05	0.05
query26	0.73	0.20	0.14
query27	0.03	0.03	0.03
query28	3.53	0.76	0.36
query29	12.53	4.08	3.21
query30	0.29	0.15	0.16
query31	2.76	0.58	0.32
query32	3.23	0.59	0.49
query33	3.22	3.18	3.31
query34	15.45	3.90	3.25
query35	3.25	3.22	3.21
query36	0.56	0.44	0.42
query37	0.09	0.07	0.06
query38	0.05	0.04	0.04
query39	0.04	0.03	0.03
query40	0.18	0.15	0.14
query41	0.08	0.03	0.03
query42	0.04	0.02	0.03
query43	0.05	0.03	0.04
Total cold run time: 96.01 s
Total hot run time: 23.94 s
```
</details>

**Comment by luwei16** [comment]
Local review result: **PASS**.

Reviewed the exact three-dot diff for head `e192f752ecf3f665bc298dd5103aadfcedb65267` against live base `befa4fda0a222f639770567ee3ac691b11ecc33f`, plus compatibility with current master `102f69734a26a97ebd88bd0bc90c627173046bb1`.

- Goal and scope: the change closes the stale-empty window for direct Cloud incremental reads by refreshing cached-empty/unknown partitions from MetaService before empty-partition pruning.
- Correctness and lifecycle: the time-based visibility fence completes before planning; cached non-empty versions are safe because partition versions are monotonic; refreshed and cached results are returned in the original partition order; final scan-node planning still fetches exact execution versions.
- Concurrency: the PR adds no shared mutable state. The existing RPC/cache path performs network work before ordered version locks and applies monotonic cache updates, with no new lock-order or race issue found.
- Parallel paths and compatibility: normal reads, non-Cloud reads, Cloud table streams, fixed-version wrappers, and time-travel semantics retain their required behavior. The PR merges cleanly with current master and introduces no protocol, persistence, or configuration compatibility change.
- Tests and CI: the three new rule tests and the existing final-version refresh test passed in remote FE UT; compile, style, regression, and performance checks passed. The separate FE coverage failure is an infrastructure Maven download failure before `build.sh --fe` started.
- Nit: the new pre-pruning MetaService lookup bypasses `SummaryProfile.incGetPartitionVersionByHasDataCount()`, so Query Profile undercounts the hasData/empty-prune reason by one RPC. Total RPC count/time and query correctness remain accurate.

Findings: Blocker 0, Major 0, Minor 0, Nit 1. The review converged in one round. This was a read-only review; no local build or test was run.


```yaml
schema: doris-repo-review/v1
status: PASS
pr: apache/doris#68138
commit: e192f752ecf3f665bc298dd5103aadfcedb65267
base: befa4fda0a222f639770567ee3ac691b11ecc33f
reviewed_at: 2026-09-18T10:52:43+08:00
reviewer: luwei16
model: gpt-5.6-sol
effort: xhigh
findings: {blocker: 0, major: 0, minor: 0, nit: 1}
rounds: 1
converged: true
```

**Review by gavinchou — DISMISSED** [review]

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PruneEmptyPartition.java [hunk 1]
```diff
@@ -67,6 +76,55 @@ public Rule build() {
         }).toRule(RuleType.PRUNE_EMPTY_PARTITION);
     }
 
+    private List<Long> selectNonEmptyPartitionIds(LogicalOlapScan scan, OlapTable table,
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PruneEmptyPartition.java:79]: 看起来应该直接更新 OlapTableWrapper 的 selectNonEmptyPartitionIds 而不是在这里写一个新的函数

**Comment by luwei16** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PruneEmptyPartition.java (hunk 1 above)
Comment (luwei16) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PruneEmptyPartition.java]: done，I moved the logic out of PruneEmptyPartition and implemented it in RowBinlogTableWrapper.selectNonEmptyPartitionIds

**Comment by hello-stephen** [comment]
<details>
<summary>TPC-H: <b>Total hot run time: 27578 ms</b></summary>

```
machine: 'aliyun_ecs.c7a.8xlarge_32C64G'
scripts: https://github.com/apache/doris/tree/master/tools/tpch-tools
Tpch sf100 test result on commit b377f1732fcba8785a1b9dd4c5a144b98114082b, data reload: false

------ Round 1 ----------------------------------
============================================
q1	17588	3918	3903	3903
q2	2257	347	297	297
q3	10003	1471	788	788
q4	4680	481	351	351
q5	7537	840	548	548
q6	186	173	138	138
q7	732	810	589	589
q8	9306	1674	1548	1548
q9	5454	4246	4230	4230
q10	6745	1629	1338	1338
q11	430	267	257	257
q12	627	407	290	290
q13	18106	2609	2009	2009
q14	264	258	233	233
q15	q16	724	713	670	670
q17	1649	1094	883	883
q18	6524	5624	5551	5551
q19	1319	1260	1074	1074
q20	491	404	266	266
q21	5898	2579	2325	2325
q22	431	355	290	290
Total cold run time: 100951 ms
Total hot run time: 27578 ms

----- Round 2, with runtime_filter_mode=off -----
============================================
q1	4203	4115	4148	4115
q2	753	588	556	556
q3	4540	4976	4437	4437
q4	2231	2292	1419	1419
q5	4273	4165	4130	4130
q6	223	170	127	127
q7	1679	1616	1408	1408
q8	2175	2488	2223	2223
q9	7558	7621	7843	7621
q10	4379	4241	3852	3852
q11	608	421	385	385
q12	760	760	523	523
q13	2442	2909	2161	2161
q14	295	305	283	283
q15	q16	716	702	636	636
q17	7794	7153	7245	7153
q18	11913	11055	11884	11055
q19	1184	1095	1078	1078
q20	2242	2283	1939	1939
q21	5824	4878	4952	4878
q22	527	472	411	411
Total cold run time: 66319 ms
Total hot run time: 60390 ms
```
</details>

**Comment by hello-stephen** [comment]
<details>
<summary>TPC-DS: <b>Total hot run time: 153943 ms</b></summary>

```
machine: 'aliyun_ecs.c7a.8xlarge_32C64G'
scripts: https://github.com/apache/doris/tree/master/tools/tpcds-tools
TPC-DS sf100 test result on commit b377f1732fcba8785a1b9dd4c5a144b98114082b, data reload: false

query5	4327	596	470	470
query6	439	198	167	167
query7	4853	519	287	287
query8	324	179	160	160
query9	8813	3969	3994	3969
query10	469	309	262	262
query11	5845	2214	2023	2023
query12	157	107	105	105
query13	1269	634	431	431
query14	6510	4591	4260	4260
query14_1	4063	4041	4010	4010
query15	202	199	185	185
query16	1014	479	453	453
query17	932	701	569	569
query18	2458	479	334	334
query19	196	189	145	145
query20	107	105	108	105
query21	225	140	119	119
query22	13116	13097	12875	12875
query23	15531	14540	13909	13909
query23_1	13975	14066	14057	14057
query24	7524	1714	1229	1229
query24_1	1251	1278	1233	1233
query25	532	454	379	379
query26	1260	319	168	168
query27	2693	547	336	336
query28	4617	1971	1969	1969
query29	1065	619	501	501
query30	307	251	211	211
query31	899	786	644	644
query32	156	101	99	99
query33	540	326	263	263
query34	1235	1122	654	654
query35	754	770	670	670
query36	783	794	763	763
query37	162	122	95	95
query38	1832	1774	1697	1697
query39	685	694	669	669
query39_1	647	672	653	653
query40	230	129	109	109
query41	73	70	68	68
query42	97	93	99	93
query43	357	371	307	307
query44	1397	714	705	705
query45	187	177	173	173
query46	1060	1169	716	716
query47	1488	1537	1395	1395
query48	392	429	311	311
query49	602	424	345	345
query50	940	355	254	254
query51	10400	10713	9964	9964
query52	84	85	73	73
query53	242	255	183	183
query54	258	217	195	195
query55	91	77	68	68
query56	250	230	215	215
query57	1528	1343	1375	1343
query58	239	206	209	206
query59	1969	2060	1833	1833
query60	277	236	226	226
query61	137	144	159	144
query62	399	320	265	265
query63	218	175	181	175
query64	2901	1078	814	814
query65	4033	3987	4003	3987
query66	1817	433	302	302
query67	20305	20053	19933	19933
query68	3420	1443	926	926
query69	413	295	265	265
query70	923	900	891	891
query71	295	238	195	195
query72	2899	2740	2249	2249
query73	808	794	431	431
query74	4634	4481	4301	4301
query75	2305	2285	1936	1936
query76	2435	1091	728	728
query77	367	399	298	298
query78	9102	9133	8558	8558
query79	1364	1188	713	713
query80	752	474	353	353
query81	491	284	234	234
query82	617	170	129	129
query83	338	275	250	250
query84	311	145	112	112
query85	1013	473	397	397
query86	380	252	232	232
query87	2007	1967	1831	1831
query88	3666	2755	2747	2747
query89	386	289	248	248
query90	1734	184	178	178
query91	172	165	123	123
query92	101	90	92	90
query93	1505	1417	843	843
query94	604	327	296	296
query95	657	442	335	335
query96	1043	809	356	356
query97	2452	2430	2300	2300
query98	195	198	184	184
query99	743	728	628	628
Total cold run time: 242109 ms
Total hot run time: 153943 ms
```
</details>

**Comment by hello-stephen** [comment]
<details>
<summary>ClickBench: <b>Total hot run time: 24.01 s</b></summary>

```
machine: 'aliyun_ecs.c7a.8xlarge_32C64G'
scripts: https://github.com/apache/doris/tree/master/tools/clickbench-tools
ClickBench test result on commit b377f1732fcba8785a1b9dd4c5a144b98114082b, data reload: false

query1	0.01	0.01	0.01
query2	0.10	0.05	0.04
query3	0.26	0.14	0.13
query4	1.61	0.14	0.14
query5	0.25	0.22	0.22
query6	1.16	0.97	0.95
query7	0.04	0.01	0.00
query8	0.05	0.03	0.04
query9	0.39	0.32	0.33
query10	0.57	0.55	0.55
query11	0.21	0.16	0.14
query12	0.19	0.15	0.15
query13	0.46	0.47	0.47
query14	0.94	0.96	0.94
query15	0.60	0.59	0.57
query16	0.31	0.32	0.32
query17	1.11	1.06	1.08
query18	0.21	0.20	0.20
query19	1.97	1.94	1.97
query20	0.02	0.01	0.01
query21	15.44	0.18	0.13
query22	5.03	0.06	0.05
query23	16.09	0.30	0.13
query24	3.00	0.40	0.31
query25	0.10	0.07	0.04
query26	0.74	0.20	0.14
query27	0.04	0.03	0.04
query28	3.63	0.77	0.36
query29	12.55	3.97	3.18
query30	0.28	0.15	0.15
query31	2.76	0.57	0.30
query32	3.24	0.59	0.49
query33	3.16	3.24	3.23
query34	15.50	3.94	3.28
query35	3.23	3.22	3.27
query36	0.57	0.43	0.42
query37	0.08	0.06	0.07
query38	0.05	0.04	0.04
query39	0.04	0.03	0.03
query40	0.17	0.15	0.14
query41	0.08	0.04	0.03
query42	0.04	0.02	0.02
query43	0.05	0.04	0.04
Total cold run time: 96.33 s
Total hot run time: 24.01 s
```
</details>

**Review by morrySnow — DISMISSED** [review]

File: fe/fe-core/src/main/java/org/apache/doris/catalog/RowBinlogTableWrapper.java [hunk 2]
```diff
@@ -79,6 +86,22 @@ public KeysType getKeysType() {
...
+            Optional<StreamReadMode> streamReadMode) {
+        if (Config.isCloudMode() && !hasFixedVisibleVersions()) {
+            // A row-binlog scan can start immediately after its target transaction becomes visible. Refresh
+            // cached-empty or unknown partitions so an older cache entry cannot prune newly visible binlog data.
+            List<CloudPartition> partitions = partitionIds.stream()
+                    .map(this::getPartition)
+                    .filter(Objects::nonNull)
+                    .map(partition -> (CloudPartition) partition)
+                    .collect(Collectors.toList());
+            return CloudPartition.selectNonEmptyPartitionIdsFromMs(partitions);
```
Comment (luwei16) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/RowBinlogTableWrapper.java:100]: [Major] Avoid routing the newly covered snapshot-version-disabled path through the two-pass cached-state classifier.

This new call does not check Config.enable_cloud_snapshot_version. The delegated helper first collects partitions whose hasDataCached() result is true, then performs a second pass to collect partitions whose result is false. If a partition returns false in the first pass and another transaction/query refreshes the shared cache to version greater than 1 between the passes, the second pass returns true. The partition is then in neither nonEmptyPartitionIds nor unknowns, is not sent to MetaService, and is silently pruned. A later ScanNode version refresh cannot restore a partition already removed from the plan, so committed binlog rows can be missed.

The helper's two-pass behavior under enable_cloud_snapshot_version=true predates this PR and is not attributed here. The regression introduced by this call is the newly covered supported false path: previously OlapTable.selectNonEmptyPartitionIds used the per-partition hasData() fallback there.

Please classify each partition exactly once in a single loop so it must enter either the known-nonempty or unknown collection, then force-refresh the unknown collection. Please also add a test covering a cached-empty partition becoming cached-nonempty between the two current checks.

**Review by luwei16 — COMMENTED** [review]
Review result for head b377f1732fcba8785a1b9dd4c5a144b98114082b: 1 Major finding, so this revision cannot receive a local PASS. The inline finding is strictly attributed to the newly covered enable_cloud_snapshot_version=false RowBinlog path; the same helper's pre-existing reachability when the option is true is not counted against this PR.

**Comment by luwei16** [comment]
Local review result: **PASS**.

Reviewed the exact three-dot diff for head `982a47dbd55001053ed936bebf4c05a0270d943e` against live base `befa4fda0a222f639770567ee3ac691b11ecc33f`, plus compatibility with current master `46e1c67122085ad624c715a96a38899fe1d8e9c6`.

- Goal and correctness: cached-empty or unknown RowBinlog partitions are refreshed from MetaService before empty-partition pruning, closing the stale-cache window that could hide newly visible committed data.
- Concurrency: the previous Major is fixed. Each partition is now classified by exactly one `hasDataCached()` call, so a concurrent false-to-true cache transition cannot make it disappear from both the known-nonempty and unknown collections. Cached-known-nonempty remains safe because visible versions are monotonic.
- RPC and locking: RPC remains outside the ordered table version locks; cache epochs, monotonic version/TSO updates, retry/error propagation, and SummaryProfile accounting retain their existing contracts.
- Lifecycle and compatibility: direct `@incr`, MOW time travel, TableBinlogFunction, fixed and unfixed Table Stream wrappers, non-Cloud mode, and `enable_cloud_snapshot_version=false` were checked. No protocol, persistence, configuration, or rolling-upgrade incompatibility was introduced. The PR merges cleanly with current master.
- Tests and CI: the deterministic `false, true` Mockito sequence plus exactly-once verification covers the old classification race. CheckStyle and lightweight GitHub checks passed on this head; FE UT and compile have not run on this head. No local build or test was run.
- Minor: the helper still returns cached-known-nonempty IDs before refreshed IDs rather than preserving candidate order, causing an unnecessary scan rewrite but no result-set error.
- Nit: both classification lists preallocate capacity for every partition, creating one avoidable empty N-capacity backing array in common all-known or all-unknown cases.

Findings: Blocker 0, Major 0, Minor 1, Nit 1. The review converged in one round.


```yaml
schema: doris-repo-review/v1
status: PASS
pr: apache/doris#68138
commit: 982a47dbd55001053ed936bebf4c05a0270d943e
base: befa4fda0a222f639770567ee3ac691b11ecc33f
reviewed_at: 2026-09-20T15:56:26+08:00
reviewer: luwei16
model: gpt-5.6-sol
effort: xhigh
findings: {blocker: 0, major: 0, minor: 1, nit: 1}
rounds: 1
converged: true
```

**Comment by luwei16** [comment]
run buildall

**Comment by hello-stephen** [comment]
<details>
<summary>TPC-H: <b>Total hot run time: 27743 ms</b></summary>

```
machine: 'aliyun_ecs.c7a.8xlarge_32C64G'
scripts: https://github.com/apache/doris/tree/master/tools/tpch-tools
Tpch sf100 test result on commit 982a47dbd55001053ed936bebf4c05a0270d943e, data reload: false

------ Round 1 ----------------------------------
============================================
q1	17784	3791	3776	3776
q2	2245	368	307	307
q3	10013	1425	781	781
q4	4685	478	351	351
q5	7530	830	562	562
q6	173	167	137	137
q7	760	798	591	591
q8	9322	1414	1376	1376
q9	5459	4162	4192	4162
q10	6848	1328	996	996
q11	439	265	249	249
q12	633	424	295	295
q13	18069	2618	2008	2008
q14	252	257	228	228
q15	q16	752	722	653	653
q17	1716	1167	1045	1045
q18	6438	5643	5567	5567
q19	1306	1203	1025	1025
q20	463	390	259	259
q21	5676	3206	3060	3060
q22	455	362	315	315
Total cold run time: 101018 ms
Total hot run time: 27743 ms

----- Round 2, with runtime_filter_mode=off -----
============================================
q1	4681	4748	4388	4388
q2	746	571	556	556
q3	4777	5207	4581	4581
q4	2227	2341	1426	1426
q5	4545	4376	4477	4376
q6	263	177	129	129
q7	1818	1652	1532	1532
q8	2312	2104	2001	2001
q9	7295	7185	7192	7185
q10	3683	3659	3188	3188
q11	530	420	332	332
q12	710	698	497	497
q13	2273	2617	2006	2006
q14	276	271	248	248
q15	q16	665	681	590	590
q17	7303	6722	6593	6593
q18	11876	11041	11754	11041
q19	1099	1020	960	960
q20	2201	2197	1914	1914
q21	4971	4072	4304	4072
q22	519	448	407	407
Total cold run time: 64770 ms
Total hot run time: 58022 ms
```
</details>

**Comment by hello-stephen** [comment]
<details>
<summary>TPC-DS: <b>Total hot run time: 152386 ms</b></summary>

```
machine: 'aliyun_ecs.c7a.8xlarge_32C64G'
scripts: https://github.com/apache/doris/tree/master/tools/tpcds-tools
TPC-DS sf100 test result on commit 982a47dbd55001053ed936bebf4c05a0270d943e, data reload: false

query5	4300	604	453	453
query6	432	208	200	200
query7	4814	529	285	285
query8	328	186	167	167
query9	8820	3961	3876	3876
query10	449	330	263	263
query11	5831	3543	3233	3233
query12	151	92	89	89
query13	1328	597	451	451
query14	6500	4507	4247	4247
query14_1	3968	3937	3928	3928
query15	203	192	184	184
query16	988	449	447	447
query17	908	676	545	545
query18	2427	466	339	339
query19	196	187	147	147
query20	80	82	79	79
query21	216	132	118	118
query22	13168	12989	12776	12776
query23	14035	13093	12567	12567
query23_1	12520	12578	12470	12470
query24	7181	1170	695	695
query24_1	715	680	734	680
query25	560	435	381	381
query26	1259	324	174	174
query27	2668	562	323	323
query28	4572	1989	1998	1989
query29	1603	725	526	526
query30	302	224	182	182
query31	893	765	632	632
query32	154	94	101	94
query33	528	321	256	256
query34	1186	1124	641	641
query35	741	751	627	627
query36	792	817	715	715
query37	141	110	94	94
query38	1826	1758	1669	1669
query39	714	700	669	669
query39_1	644	653	654	653
query40	220	126	107	107
query41	77	69	67	67
query42	100	94	90	90
query43	340	347	307	307
query44	1357	706	717	706
query45	232	173	164	164
query46	1093	1179	735	735
query47	1485	1470	1391	1391
query48	410	423	303	303
query49	572	409	289	289
query50	1002	358	252	252
query51	10579	10604	10449	10449
query52	86	84	77	77
query53	234	253	182	182
query54	246	200	191	191
query55	77	71	67	67
query56	217	202	224	202
query57	1464	1404	1302	1302
query58	285	261	250	250
query59	2004	2068	1863	1863
query60	275	243	219	219
query61	137	145	140	140
query62	395	317	268	268
query63	214	175	178	175
query64	2774	969	791	791
query65	3483	3424	3419	3419
query66	1816	405	362	362
query67	20031	19775	19801	19775
query68	3161	1441	870	870
query69	429	300	254	254
query70	916	817	806	806
query71	290	232	205	205
query72	2653	2496	2239	2239
query73	847	739	442	442
query74	4609	4500	4299	4299
query75	2287	2263	1941	1941
query76	2299	1089	734	734
query77	347	391	296	296
query78	9217	9061	8468	8468
query79	1432	1211	772	772
query80	1263	450	357	357
query81	615	324	280	280
query82	673	165	123	123
query83	293	220	204	204
query84	316	141	109	109
query85	862	472	384	384
query86	396	223	211	211
query87	2008	1966	1837	1837
query88	3626	2705	2684	2684
query89	379	287	246	246
query90	1912	179	177	177
query91	174	156	128	128
query92	94	81	88	81
query93	1638	1480	853	853
query94	744	334	295	295
query95	663	367	341	341
query96	987	774	371	371
query97	2422	2435	2341	2341
query98	157	151	143	143
query99	730	729	622	622
Total cold run time: 237349 ms
Total hot run time: 152386 ms
```
</details>

**Comment by hello-stephen** [comment]
<details>
<summary>ClickBench: <b>Total hot run time: 23.99 s</b></summary>

```
machine: 'aliyun_ecs.c7a.8xlarge_32C64G'
scripts: https://github.com/apache/doris/tree/master/tools/clickbench-tools
ClickBench test result on commit 982a47dbd55001053ed936bebf4c05a0270d943e, data reload: false

query1	0.00	0.00	0.00
query2	0.09	0.04	0.04
query3	0.26	0.14	0.13
query4	1.61	0.14	0.14
query5	0.25	0.21	0.21
query6	1.15	0.93	0.92
query7	0.03	0.01	0.00
query8	0.05	0.04	0.03
query9	0.38	0.34	0.34
query10	0.54	0.59	0.58
query11	0.20	0.15	0.14
query12	0.19	0.15	0.16
query13	0.46	0.47	0.47
query14	0.96	0.95	0.93
query15	0.62	0.61	0.58
query16	0.32	0.33	0.30
query17	1.08	1.06	1.07
query18	0.22	0.20	0.20
query19	2.03	1.95	1.96
query20	0.02	0.01	0.01
query21	15.44	0.22	0.14
query22	4.90	0.05	0.06
query23	16.12	0.31	0.12
query24	2.88	0.43	0.33
query25	0.12	0.05	0.04
query26	0.75	0.20	0.14
query27	0.05	0.04	0.03
query28	3.51	0.73	0.36
query29	12.51	4.06	3.21
query30	0.29	0.15	0.16
query31	2.77	0.55	0.32
query32	3.23	0.59	0.49
query33	3.22	3.14	3.15
query34	15.50	3.96	3.32
query35	3.20	3.20	3.23
query36	0.54	0.44	0.41
query37	0.09	0.06	0.06
query38	0.05	0.04	0.04
query39	0.04	0.03	0.03
query40	0.18	0.15	0.15
query41	0.08	0.03	0.02
query42	0.04	0.04	0.03
query43	0.04	0.03	0.04
Total cold run time: 96.01 s
Total hot run time: 23.99 s
```
</details>

**Comment by luwei16** [comment]
run cloud_p0

**Comment by luwei16** [comment]
run cloud_p0

**Comment by luwei16** [comment]
run nonConcurrent

**Comment by hello-stephen** [comment]
# FE Regression Coverage Report
Increment line coverage `  90.48% (19/21)` :tada:
[Increment coverage report](http://coverage.selectdb-in.cc/coverage/68138_982a47dbd55001053ed936bebf4c05a0270d943e_merge_fe/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/68138_982a47dbd55001053ed936bebf4c05a0270d943e_merge_fe/report/index.html)

**Review by gavinchou — APPROVED** [review]

**Review by morrySnow — APPROVED** [review]
